File: frontend/app/models/item.js
import Ember from 'ember';
import DS from 'ember-data';
import moment from 'moment';

export default DS.Model.extend({
  i18n: Ember.inject.service(),

  biblio: DS.belongsTo('biblio'),
  sublocation: DS.belongsTo('sublocation'),
  canBeOrdered: DS.attr('boolean'),
  canBeQueued: DS.attr('boolean'),
  itemType: DS.attr('string'),
  itemCallNumber: DS.attr('string'),
  copyNumber: DS.attr('string'),
  barcode: DS.attr('string'),
  status: DS.attr('string'),
  statusLimitation: DS.attr('string'),
  dueDate: DS.attr('string'),
  notForLoan: DS.attr('string'),
  isReserved: DS.attr('boolean'),
  isAvailible: DS.attr('boolean'),

  sublocationText: Ember.computed('sublocation.isOpenLoc', 'sublocation.name', 'itemCallNumber', function() {
    const dictionary = this.get('i18n');
    if (this.get('sublocation.isOpenLoc')) {
      if (this.get('itemCallNumber')) {
        return this.get('sublocation.name') + ', ' + this.get('itemCallNumber');
      } else if (this.get('biblio.biblioCallNumber')) {
        return this.get('sublocation.name') + ', ' + this.get('biblio.biblioCallNumber');
      } else {
        return this.get('sublocation.name');
      }
    } else {
      return dictionary.t('components.item-table.must_be_ordered');
    }
  }),

  statusLimitationText: Ember.computed('statusLimitation', function() {
    const dictionary = this.get('i18n');
    if (this.get('statusLimitation') === 'NOT_FOR_HOME_LOAN') {
      return dictionary.t('components.item-table.not_for_home_loan');
    } else if (this.get('statusLimitation') === 'READING_ROOM_ONLY') {
      return dictionary.t('components.item-table.reading_room_only');
    } else {
      return '';
    }
  }),

  statusText: Ember.computed('status', 'dueDate', function() {
    const dictionary = this.get('i18n');

    if (this.get('status') === 'LOANED') {
      return dictionary.t('components.item-table.loaned') + ' ' + this.get('dueDate').slice(0, -8); // + moment(this.get('dueDate')).format("YYYY-MM-DD");
    } else if (this.get('status') === 'RESERVED') {
      return dictionary.t('components.item-table.reserved');
    } else if (this.get('status') === 'WAITING') {
      return dictionary.t('components.item-table.waiting');
    } else if (this.get('status') === 'IN_TRANSIT') {
      return dictionary.t('components.item-table.in_transit');
    } else if (this.get('status') === 'DELAYED') {
      return dictionary.t('components.item-table.delayed');
    } else if (this.get('status') === 'DURING_ACQUISITION') {
      return dictionary.t('components.item-table.under_acquisition');
    } else if (this.get('status') === 'NOT_IN_PLACE') {
      return dictionary.t('components.item-table.not_in_place');
    } else if (this.get('status') === 'AVAILABLE') {
      return dictionary.t('components.item-table.available');
    } else {
      return dictionary.t('components.item-table.unknown');
    }
  }),

  actions: {
    setItemToOrder(item) {
      this.get('setItemToOrder')(item);
    }
  }
});

File: app/models/item.rb
class Item
  include ActiveModel::Serialization
  include ActiveModel::Validations

  attr_accessor :id, :biblio_id, :sublocation_id, :item_type, :barcode, :item_call_number,
                :copy_number, :public_notes, :due_date, :lost, :restricted, :not_for_loan, :is_reserved,
                :withdrawn, :in_transit, :location_id, :location_name_sv,:sublocation_open_loc

  attr_writer :found

  def initialize biblio_id:, rawdata:, has_item_level_queue: false
    @biblio_id = biblio_id
    @has_item_level_queue = has_item_level_queue
    @is_reserved = false
    parse_rawdata(rawdata)
    sublocation = Sublocation.find_by_id(@sublocation_id)
    @sublocation_open_loc = sublocation_open_loc?
    @sublocation_name_sv = sublocation.name_sv
    @sublocation_name_en = sublocation.name_en
    @location_id = sublocation.location_id
    location = Location.find_by_id(@location_id)
    @location_name_sv = location.name_sv
    @location_name_en = location.name_en
  end

  def as_json options = {}
    super(except: ["found"]).merge({can_be_ordered: can_be_ordered,
      can_be_queued: can_be_queued,
      status: status,
      status_limitation: status_limitation,
      is_availible: is_availible
      }).compact
  end

  def can_be_borrowed
    # TODO Extend with more rules
    !item_type_ref?
  end

  def is_availible
    return false if reserved?
    return false if checked_out?
    return false if not_in_place?
    return false if during_acquisition?
    return false if in_transit?
    return true
  end


  def can_be_ordered
    # TODO Extend with more rules

    return false unless @item_type
    return false if item_type_ref?
# CORONA: Make Kursbok orderable
#    return false if item_type_kursbok?
    return false if checked_out?
    return false if reserved?
    return false if lost?
    return false if during_acquisition?
    return false if not_in_place?
    return false if in_transit?
    return false unless sublocation_paging_loc?
    return true
  end

  def can_be_queued
    @has_item_level_queue && is_available_for_queue
  end

  def is_available_for_queue
    return false if item_type_ref?
    return false if restricted?
    return false unless checked_out? || reserved? || during_acquisition? || not_in_place? || in_transit?
    return true
  end

  def sublocation_paging_loc?
    Sublocation.find_by_id(@sublocation_id).is_paging_loc == '1'
  end

  def sublocation_open_loc?
    Sublocation.find_by_id(@sublocation_id).is_open_loc == '1'
  end
  def lost?
    @lost != '0'
  end

  def checked_out?
    @due_date.present?
  end

  def reserved?
    @is_reserved
  end

  def restricted?
    # ['1', '2', '5', '6'].include?(@restricted)
    !(@restricted == '0' || @restricted.nil?)
  end

  def item_type_ref?
    @item_type == '7'
  end

  def item_type_kursbok?
    @item_type == '2'
  end

  def masked?
    ['1', '2', '3'].include?(@withdrawn) || ['1', '5'].include?(@lost)
  end

  def not_in_place?
    @lost == '4'
  end

  def during_acquisition?
    @withdrawn == '4'
  end

  def in_transit?
    @in_transit == '1' || (@currentlocation_id != @location_id)
  end

  def status_limitation
    return "NOT_FOR_HOME_LOAN" if item_type_ref?
    return "READING_ROOM_ONLY" if @not_for_loan == '-3'
    return "LOAN_IN_HOUSE_ONLY" if ['3', '4', '5', '6', '7', '8'].include?(@restricted)
  end

  def status
    return "LOANED" if @due_date.present? && Date.parse(@due_date) >= Date.today
    return "RESERVED" if (@is_reserved && @due_date.blank?) || @found == "W" || @found == "T"
    return "DELAYED" if (@due_date.present? && Date.parse(@due_date) < Date.today) || ['2', '3', '6'].include?(@lost)
    return "IN_TRANSIT" if in_transit?
    return "NOT_IN_PLACE" if not_in_place?
    return "DURING_ACQUISITION" if during_acquisition?
    return "AVAILABLE"
  end

  def parse_rawdata rawdata
    rawdata["itemnumber"].present? ? @id = rawdata["itemnumber"] : @id = nil
    rawdata["location"].present? ? @sublocation_id = rawdata["location"] : @sublocation_id = nil
    rawdata["itype"].present? ? @item_type = rawdata["itype"].to_s : @item_type = nil
    rawdata["barcode"].present? ? @barcode = rawdata["barcode"] : @barcode = nil
    rawdata["itemcallnumber"].present? ? @item_call_number = rawdata["itemcallnumber"] : @item_call_number = nil
    rawdata["enumchron"].present? ? @copy_number = rawdata["enumchron"] : @copy_number = nil
    rawdata["itemnotes"].present? ? @public_notes = rawdata["itemnotes"] : @public_notes = nil
    rawdata["itemlost"].present? ? @lost = rawdata["itemlost"].to_s : @lost = nil
    rawdata["restricted"].present? ? @restricted = rawdata["restricted"].to_s : @restricted = nil
    rawdata["notforloan"].present? ? @not_for_loan = rawdata["notforloan"].to_s : @not_for_loan = nil
    rawdata["withdrawn"].present? ? @withdrawn = rawdata["withdrawn"].to_s : @withdrawn = nil
    rawdata["datedue"].present? ? @due_date = rawdata["datedue"] : @due_date = nil
    rawdata["in_transit"].present? ? @in_transit = rawdata["in_transit"] : @in_transit = nil
    rawdata["holdingbranch"].present? ? @currentlocation_id = rawdata["holdingbranch"] : @currentlocation_id = nil
  end
end

